applications/minotari_app_grpc/src/authentication/basic_auth.rs

```diff
@@ -20,24 +20,66 @@
 //  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 //  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-use std::{borrow::Cow, ops::Deref, string::FromUtf8Error};
+use std::{
+    borrow::Cow,
+    cmp::min,
+    ops::{BitAnd, Deref},
+    string::FromUtf8Error,
+};
 
 use argon2::{password_hash::Encoding, Argon2, PasswordHash, PasswordVerifier};
-use tari_utilities::SafePassword;
+use rand::RngCore;
+use subtle::{Choice, ConstantTimeEq};
+use tari_utilities::{ByteArray, SafePassword};
 use tonic::metadata::{errors::InvalidMetadataValue, Ascii, MetadataValue};
 use zeroize::{Zeroize, Zeroizing};
 
+use crate::authentication::salted_password::create_salted_hashed_password;
+
+const MAX_USERNAME_LEN: usize = 256;
+
 /// Implements [RFC 2617](https://www.ietf.org/rfc/rfc2617.txt#:~:text=The%20%22basic%22%20authentication%20scheme%20is,other%20realms%20on%20that%20server.)
 /// Represents the username and password contained within a Authenticate header.
 #[derive(Debug)]
 pub struct BasicAuthCredentials {
-    pub user_name: String,
-    pub password: SafePassword,
+    /// The username bytes length
+    pub user_name_bytes_length: usize,
+    /// The username in bytes representation for constant time comparison
+    pub user_name_bytes: [u8; MAX_USERNAME_LEN],
+    /// The hashed password
+    pub phc_password_hash: SafePassword,
+    /// Random bytes to help with constant time username comparison
+    pub random_bytes: [u8; MAX_USERNAME_LEN],
 }
 
 impl BasicAuthCredentials {
-    pub fn new(user_name: String, password: SafePassword) -> Self {
-        Self { user_name, password }
+    /// Creates a new `Credentials` instance from a username and password (PHC string bytes).
+    pub fn new(user_name: String, phc_password_hash: SafePassword) -> Result<Self, BasicAuthError> {
+        // Validate the username is well formed
+        if user_name.as_bytes().len() > MAX_USERNAME_LEN {
+            return Err(BasicAuthError::InvalidUsername);
+        }
+        // Validate the password is a well formed byte representation of a PHC string
+        let bytes = phc_password_hash.reveal().to_vec();
+        let _parse_result = PasswordHash::parse(&String::from_utf8(bytes)?, Encoding::B64)?;
+        // Random bytes are used for constant time username comparison to ensure that the compiler does not do any
+        // funny optimizations and to ensure that comparison for the same username for every new credentials instance
+        // forces a different bitwise comparison.
```

Comment:
I still disagree that the use of random bytes here is necessary, as it could only become relevant in the event that `subtle` fails, and I'm not a fan of requiring the RNG unless necessary. That being said, I get the defense-in-depth approach here and wouldn't block review on it.